### DISCUSSION TITLE
4.3.51-iOS版本同时播放多个特定的特效会产生崩溃

### DISCUSSION THREAD

**Discussion by LUJYM** [Tencent/libpag#2272] [Q&A] [answered]
系统：iOS
环境：4.3.51静态库版本
描述：同时播放2个相同特定的特效会产生崩溃
文件：在一个页面里面同时创建两个pagview去播放aaa.pag数秒后就会崩溃
[aaaa.pag.zip](https://github.com/Tencent/libpag/files/15125226/aaaa.pag.zip)
demo：
[libpag.zip](https://github.com/Tencent/libpag/files/15140490/libpag.zip)
![image](https://github.com/Tencent/libpag/assets/24284466/7cf1ed3f-8724-4a21-998e-9f1577fde72c)
崩溃堆栈如下：
`* thread #15, stop reason = EXC_BAD_ACCESS (code=1, address=0x10)
  * frame #0: 0x00000001004c7700 pag-ios`pag::AnimatableProperty<float>::getValueAt(long long) + 188
    frame #1: 0x00000001004b76d4 pag-ios`pag::FastBlurEffect::visibleAt(long long) const + 24
    frame #2: 0x0000000100560368 pag-ios`pag::FilterRenderer::MakeFilterList(pag::FilterModifier const*) + 312
    frame #3: 0x00000001005611cc pag-ios`pag::FilterRenderer::DrawWithFilter(tgfx::Canvas*, pag::RenderCache*, pag::FilterModifier const*, std::__1::shared_ptr<pag::Graphic>) + 88
    frame #4: 0x0000000100539d4c pag-ios`pag::FilterModifier::applyToGraphic(tgfx::Canvas*, pag::RenderCache*, std::__1::shared_ptr<pag::Graphic>) const + 72
    frame #5: 0x0000000100547dec pag-ios`pag::ModifierGraphic::draw(tgfx::Canvas*, pag::RenderCache*) const + 100
    frame #6: 0x00000001005473e0 pag-ios`pag::MatrixGraphic::draw(tgfx::Canvas*, pag::RenderCache*) const + 72
    frame #7: 0x0000000100547910 pag-ios`pag::LayerGraphic::draw(tgfx::Canvas*, pag::RenderCache*) const + 68
    frame #8: 0x000000010054884c pag-ios`pag::ClipModifier::applyToGraphic(tgfx::Canvas*, pag::RenderCache*, std::__1::shared_ptr<pag::Graphic>) const + 76
    frame #9: 0x0000000100547dec pag-ios`pag::ModifierGraphic::draw(tgfx::Canvas*, pag::RenderCache*) const + 100
    frame #10: 0x00000001005473e0 pag-ios`pag::MatrixGraphic::draw(tgfx::Canvas*, pag::RenderCache*) const + 72
    frame #11: 0x000000010054884c pag-ios`pag::ClipModifier::applyToGraphic(tgfx::Canvas*, pag::RenderCache*, std::__1::shared_ptr<pag::Graphic>) const + 76
    frame #12: 0x0000000100547dec pag-ios`pag::ModifierGraphic::draw(tgfx::Canvas*, pag::RenderCache*) const + 100
    frame #13: 0x0000000100519be4 pag-ios`pag::PAGSurface::draw(pag::RenderCache*, std::__1::shared_ptr<pag::Graphic>, pag::BackendSemaphore*, bool) + 448
    frame #14: 0x00000001005182bc pag-ios`pag::PAGPlayer::flushInternal(pag::BackendSemaphore*) + 148
    frame #15: 0x00000001005181e4 pag-ios`pag::PAGPlayer::flush() + 140
    frame #16: 0x00000001004fe6f0 pag-ios`pag::AnimatorListener::onAnimationUpdate(pag::PAGAnimator*) + 96
    frame #17: 0x0000000100512d94 pag-ios`pag::PAGAnimator::onFlush(bool) + 84
    frame #18: 0x00000001005139f8 pag-ios`std::__1::__function::__func<pag::PAGAnimator::doUpdate(bool)::$_0, std::__1::allocator<pag::PAGAnimator::doUpdate(bool)::$_0>, void ()>::operator()() + 52
    frame #19: 0x00000001005cd4b4 pag-ios`tgfx::Task::execute() + 36
    frame #20: 0x00000001005cd848 pag-ios`tgfx::TaskGroup::RunLoop(tgfx::TaskGroup*) + 52
    frame #21: 0x00000001005ce270 pag-ios`void* std::__1::__thread_proxy[abi:v15006]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void (*)(tgfx::TaskGroup*), tgfx::TaskGroup*>>(void*) + 52
    frame #22: 0x00000001f18886cc libsystem_pthread.dylib`_pthread_start + 148`

**Comment by kevingpqi123** [comment]
麻烦确认下使用的方式，播放相同的 pag 文件通过的是 PAGView setPath 还是 setComposition，如果通过的是 setComposition，需要确保加载的不是同一个 PAGComposition 对象， 此时每一个 PAGComposition 对象都需要通过 PAGFile Load 的方式获取

另外，如果一个页面需要播放两个相同的 pag 文件，从性能的角度出发，我们推荐使用 pag 的组合模式，创建一个空的 PAGComposition，将两个 PAGFile 通过 addLayer 的方式添加进去，具体可以看下官方 demo 中的范例

**Reply by LUJYM** [reply]
使用方式是setComposition，加载的是不同的PAGComposition对象，另外播放多个的pag文件是业务有这种场景（比如麦位特效这种），崩溃的这个在提的dis上面是有必现的demo的哈，大佬可以看下

**Reply by kevingpqi123** [reply]
已定位到原因，本周会提供一个修复版本，感谢反馈

**Reply by kevingpqi123** [reply, marked as answer]
可以升级至 4.3.54, 看问题是否解决

**Reply by LUJYM** [reply]
用了4.3.56版本测试，此崩溃问题已经修复👍🏻

**Comment by LUJYM** [comment]
> 可以升级至 4.3.54, 看问题是否解决
